"percentage" mode - category subtotal calc errors

The category subtotal (and possibly course grade? my math skills are not up to the challenge of figuring that out) don't seem to be reflected correctly when in "Percentage" mode.

Note: for the following, I'm using categories & weighting.

![image](https://cloud.githubusercontent.com/assets/11447474/15904527/5de7cbaa-2d7f-11e6-8291-1eea61c6e452.png)

In the image above, you can see that there are three items with scores and different relative item weights (one worth 300, one worth 100, and one worth 200).

For Julio tester, a cat subtotal score of 31.67% is reported. However, they should have a 53.33%. Maybe I'm mistaken, but I think the calculation for this should be:

( (.4 * 300) + (1.0 * 100) + (.5 * 200) ) / (300 + 100 + 200) = .5333

Likewise, Taifa tester is reporting a cat subtotal score of 15%. This should be 30%, with the calculation for this being:

( (0 * 300) + (.75 * 200) ) / (300 + 200) = .3

Category subtotal doesn't take into account percentages. This needs to be added. Likewise for letter grades, otherwise you get this error:
````
aused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.apache.wicket.session.DefaultPageFactory.newPage(DefaultPageFactory.java:175)
	... 61 more
Caused by: java.lang.NumberFormatException: For input string: "A-"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at java.lang.Double.valueOf(Double.java:502)
	at org.apache.commons.lang.math.NumberUtils.createDouble(NumberUtils.java:698)
	at org.sakaiproject.component.gradebook.GradebookServiceHibernateImpl.calculateCategoryScore(GradebookServiceHibernateImpl.java:3041)
````

--- a/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java
+++ b/edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java
@@ -3031,15 +3031,29 @@ public class GradebookServiceHibernateImpl extends BaseHibernateManager implemen
 	public Double calculateCategoryScore(Object gradebook, String studentUuid, CategoryDefinition category, final List<org.sakaiproject.service.gradebook.shared.Assignment> viewableAssignments, Map<Long,String> gradeMap) {
 		
 		Gradebook gb = (Gradebook) gradebook;
-				
+		
+		// used for translating letter grades
+		final Map<String, Double> gradingSchema = gb.getSelectedGradeMapping().getGradeMap();
+		
 		//collect the data and turn it into a list of AssignmentGradeRecords
 		//this is the info that is compatible with both applyDropScores and the calculateCategoryScore method
 		List<AssignmentGradeRecord> gradeRecords = new ArrayList<>();
 		for(org.sakaiproject.service.gradebook.shared.Assignment assignment: viewableAssignments) {
 			
 			Long assignmentId = assignment.getId();
-			Double grade = NumberUtils.createDouble(gradeMap.get(assignmentId));
 			
+			String rawGrade = gradeMap.get(assignmentId);
+			Double pointsPossible = assignment.getPoints();
+			Double grade = null;
+			
+			//determine the grade we should be using depending on the grading type
+			if (gb.getGrade_type() == GradebookService.GRADE_TYPE_PERCENTAGE) {
+				grade = calculateEquivalentPointValueForPercent(pointsPossible, NumberUtils.createDouble(rawGrade));
+			}
+			if (gb.getGrade_type() == GradebookService.GRADE_TYPE_LETTER){
+				grade = gradingSchema.get(rawGrade);
+			}
+									
 			//recreate the category (required fields only)
 			Category c = new Category();
 			c.setId(category.getId());
@@ -3121,7 +3135,7 @@ public class GradebookServiceHibernateImpl extends BaseHibernateManager implemen
 			Assignment assignment = gradeRecord.getAssignment();
 						
 			//check category ids match, otherwise skip
-			if(assignment.getCategory() != null && categoryId != null && categoryId.longValue() != assignment.getCategory().getId().longValue()){
+			if(assignment.getCategory() != null && categoryId.longValue() != assignment.getCategory().getId().longValue()){
 				continue;
 			}
 						

